src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -651,20 +686,39 @@ private void ParseQuoted(HtmlSymbolType type)
 
         private bool StartTag(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, IDisposable tagBlockWrapper)
         {
-            // If we're at text, it's the name, otherwise the name is ""
             HtmlSymbol tagName;
-            if (At(HtmlSymbolType.Text))
+            HtmlSymbol bangSymbol = null;
+            HtmlSymbol potentialTagNameSymbol;
+
+            if (At(HtmlSymbolType.Bang))
+            {
+                bangSymbol = CurrentSymbol;
+
+                potentialTagNameSymbol = Lookahead();
```

Comment:
nit: suggest making the parameter to `Lookahead()` required.  otherwise reader needs to hit `F12` to see whether it will return the current, next, or some random symbol.